### PR TITLE
Increase use of parse_map in material models

### PR DESCRIPTION
**Pull request by bobmyhill**
This PR replaces use of the "possibly_extend_from_1_to_N" function with the more flexible "parse_map_to_double_array" in all material models apart from steinberger.

This is a small step towards unifying code functionality across material models.

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).
* [x] I have tested my new feature locally to ensure it is correct.

### CODE REVIEW

**Review by MFraters — APPROVED** [review]
The changes look reasonable to me. Is there any reason you explicitly left the Steinberger material model out? Is there an issue there?

**Comment by bobmyhill** [comment]
@MFraters: steinberger (and melt_boukare) use compositional fields in a slightly different way to the other material models. In these models, the number of important `equation of state parameter fields n_p` does not correspond with the number of tracked `chemical component fields n_c`. 

In the case of melt_boukare the modification was easy to make (because n_p = the number of chemical endmembers, whose names are defined in the Material Model). For steinberger the EoS properties are provided in a lookup table, so `n_p = 0`, and the Material model parameter fields are instead reserved for different thermal conductivity domains, defined according to depth. The different depth regions are never given names, so `parse_map_to_double_array` is not a natural function to use. I also don't see anyone using phases in this material model (much better to add thermal conductivities to the lookup).

TL;DR: There's no issue with steinberger, it's just more natural to keep the current function calls for that particular model.

**Review by MFraters — APPROVED** [review]
Thanks for the explanation. Looks good to me and I see no other comments, so I am going to merge it. You are really on a roll with pull requests!
